runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validator.go

```diff
@@ -0,0 +1,372 @@
+
+	underlyingValidator := NewContractUpdateValidator(location, contractName, provider, oldProgram, newProgram)
+
+	return &CadenceV042ToV1ContractUpdateValidator{
+		underlyingUpdateValidator: underlyingValidator,
+		newElaborations:           newElaborations,
+	}
+}
+
+var _ UpdateValidator = &CadenceV042ToV1ContractUpdateValidator{}
+
+func (validator *CadenceV042ToV1ContractUpdateValidator) getCurrentDeclaration() ast.Declaration {
+	return validator.underlyingUpdateValidator.getCurrentDeclaration()
+}
+
+func (validator *CadenceV042ToV1ContractUpdateValidator) setCurrentDeclaration(decl ast.Declaration) {
+	validator.underlyingUpdateValidator.setCurrentDeclaration(decl)
+}
+
+func (validator *CadenceV042ToV1ContractUpdateValidator) getAccountContractNames(address common.Address) ([]string, error) {
+	return validator.underlyingUpdateValidator.accountContractNamesProvider.GetAccountContractNames(address)
+}
+
+// Validate validates the contract update, and returns an error if it is an invalid update.
+func (validator *CadenceV042ToV1ContractUpdateValidator) Validate() error {
+	underlyingValidator := validator.underlyingUpdateValidator
+
+	oldRootDecl := getRootDeclaration(validator, underlyingValidator.oldProgram)
+	if underlyingValidator.hasErrors() {
+		return underlyingValidator.getContractUpdateError()
+	}
+
+	newRootDecl := getRootDeclaration(validator, underlyingValidator.newProgram)
+	if underlyingValidator.hasErrors() {
+		return underlyingValidator.getContractUpdateError()
+	}
+
+	validator.TypeComparator.RootDeclIdentifier = newRootDecl.DeclarationIdentifier()
+	validator.TypeComparator.expectedIdentifierImportLocations = collectImports(validator, underlyingValidator.oldProgram)
+	validator.TypeComparator.foundIdentifierImportLocations = collectImports(validator, underlyingValidator.newProgram)
+
+	checkDeclarationUpdatability(validator, oldRootDecl, newRootDecl)
+
+	if underlyingValidator.hasErrors() {
+		return underlyingValidator.getContractUpdateError()
+	}
+
+	return nil
+}
+
+func (validator *CadenceV042ToV1ContractUpdateValidator) report(err error) {
+	validator.underlyingUpdateValidator.report(err)
+}
+
+func (validator *CadenceV042ToV1ContractUpdateValidator) idAndLocationOfQualifiedType(typ *ast.NominalType) (common.TypeID, common.Location) {
+
+	qualifiedString := typ.String()
+
+	// working under the assumption that any new program we are validating already typechecks,
+	// any nominal type must fall into one of three cases:
+	// 1) type qualified by an import (e.g `C.R` where `C` is an imported type)
+	// 2) type qualified by the root declaration (e.g `C.R` where `C` is the root contract or contract interface of the new contract)
+	// 3) unqualified type (e.g. `R`, but declared inside `C`)
+	//
+	// in case 3, we prepend the root declaration identifier with a `.` to the type's string to get its qualified name,
+	// and in 1 and 2 we don't need to do anything
+	typIdentifier := typ.Identifier.Identifier
```

Comment:
Does this need to handle `ast.NominalType`'s `NestedIdentifiers`?

Reply:
In case that's the case, let's improve it in a follow-up PR and get this PR in, as it blocks a few other follow-up PRs already.

Reply:
not sure. I didn't run into any issues when running this agains the core contracts. But it does looks like we may need to handle `NestedIdentifiers` (or atleast skip if there are already nested identifiers?)

+1 for doing the improvements separately. Need one more approval to merge the PR.

Reply:
👍 Could you please add a TODO to #2642?

Reply:
https://github.com/onflow/cadence/issues/3127